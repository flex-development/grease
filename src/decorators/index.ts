/**
 * @file Entry Point - Decorators
 * @module grease/decorators
 */

export { default as IsManifestId } from './is-manifest-id.decorator'
export { default as IsNilable } from './is-nilable.decorator'
export { default as IsNullable } from './is-nullable.decorator'
export { default as IsOptional } from './is-optional.decorator'
export { default as IsReleaseVersion } from './is-release-version.decorator'
