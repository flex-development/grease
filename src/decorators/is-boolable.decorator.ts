/**
 * @file Decorators - IsBoolable
 * @module grease/decorators/IsBoolable
 */

import {
  cast,
  isBoolean,
  type DecoratorTarget,
  type Objectify,
  type OwnPropertyKey,
  type PropertyDecorator
} from '@flex-development/tutils'
import {
  ValidationTypes,
  getMetadataStorage,
  type ValidationOptions
} from 'class-validator'
import {
  ValidationMetadata
} from 'class-validator/esm2015/metadata/ValidationMetadata'

/**
 * Ignore all validators if a property value is a boolean.
 *
 * @decorator
 *
 * @param {ValidationOptions} [options] - Validation options
 * @return {PropertyDecorator} Property decorator
 */
const IsBoolable = (options: ValidationOptions = {}): PropertyDecorator => {
  return (target: DecoratorTarget, name: OwnPropertyKey): void => {
    getMetadataStorage().addValidationMetadata(
      new ValidationMetadata({
        constraints: [(obj: Objectify<any>) => !isBoolean(obj[name])],
        propertyName: cast(name),
        target: target.constructor,
        type: ValidationTypes.CONDITIONAL_VALIDATION,
        validationOptions: options
      })
    )
  }
}

export default IsBoolable
