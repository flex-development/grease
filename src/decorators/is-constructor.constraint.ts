/**
 * @file Decorators - IsConstructorConstraint
 * @module grease/decorators/IsConstructorConstraint
 */

import {
  cast,
  get,
  isFunction,
  template,
  type Constructor,
  type Optional
} from '@flex-development/tutils'
import {
  ValidatorConstraint,
  buildMessage,
  type ValidatorConstraintInterface as IValidatorConstraint,
  type ValidationArguments,
  type ValidationOptions
} from 'class-validator'

/**
 * Function validator.
 *
 * @class
 * @implements {IValidatorConstraint}
 */
@ValidatorConstraint(IsConstructorConstraint.options)
class IsConstructorConstraint implements IValidatorConstraint {
  /**
   * Validator constraint options.
   *
   * @public
   * @static
   * @readonly
   * @member {{ async: boolean; name: string }} options
   */
  public static readonly options: { async: boolean; name: string } = {
    async: false,
    name: 'isConstructor'
  }

  /**
   * Get the default validation failure message.
   *
   * @see {@linkcode ValidationArguments}
   *
   * @public
   *
   * @param {ValidationArguments} args - Validation arguments
   * @return {string} Default validation failure message
   */
  public defaultMessage(args: ValidationArguments): string {
    return buildMessage(
      prefix => template('{prefix}$property must be a constructor', { prefix }),
      cast<Optional<ValidationOptions>>(get(args, 'constraints.0'))
    )()
  }

  /**
   * Check if `value` is a constructor.
   *
   * @public
   *
   * @param {unknown} value - Value to check
   * @return {value is Constructor<any>} `true` if value is a constructor
   */
  public validate(value: unknown): value is Constructor<any> {
    if (isFunction(value)) {
      try {
        new new Proxy<Constructor<any>>(cast(value), {
          construct() {
            return {}
          }
        })()
        return true
      } catch {
        return false
      }
    }

    return false
  }
}

export default IsConstructorConstraint
