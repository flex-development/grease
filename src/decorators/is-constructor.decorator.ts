/**
 * @file Decorators - IsConstructor
 * @module grease/decorators/IsConstructor
 */

import type { PropertyDecorator } from '@flex-development/tutils'
import { ValidateBy, type ValidationOptions } from 'class-validator'
import validator from './is-constructor.constraint'

/**
 * Check if a property value is a constructor.
 *
 * @decorator
 *
 * @param {ValidationOptions} [options] - Validation options
 * @return {PropertyDecorator} Property decorator
 */
const IsConstructor = (options: ValidationOptions = {}): PropertyDecorator => {
  return ValidateBy({
    async: validator.options.async,
    constraints: [options],
    name: validator.options.name,
    validator
  }, options)
}

export default IsConstructor
