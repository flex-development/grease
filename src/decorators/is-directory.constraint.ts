/**
 * @file Decorators - IsDirectoryConstraint
 * @module grease/decorators/IsDirectoryConstraint
 */

import { isDirectory } from '@flex-development/mlly'
import { isAbsolute } from '@flex-development/pathe'
import {
  cast,
  get,
  isString,
  template,
  type Optional
} from '@flex-development/tutils'
import {
  ValidatorConstraint,
  buildMessage,
  type ValidatorConstraintInterface as IValidatorConstraint,
  type ValidationArguments,
  type ValidationOptions
} from 'class-validator'

/**
 * Directory path validator.
 *
 * @class
 * @implements {IValidatorConstraint}
 */
@ValidatorConstraint(IsDirectoryConstraint.options)
class IsDirectoryConstraint implements IValidatorConstraint {
  /**
   * Validator constraint options.
   *
   * @public
   * @static
   * @readonly
   * @member {{ async: boolean; name: string }} options
   */
  public static readonly options: { async: boolean; name: string } = {
    async: false,
    name: isDirectory.name
  }

  /**
   * Get the default validation failure message.
   *
   * @see {@linkcode ValidationArguments}
   *
   * @public
   *
   * @param {ValidationArguments} args - Validation arguments
   * @return {string} Default validation failure message
   */
  public defaultMessage(args: ValidationArguments): string {
    /**
     * Default message template.
     *
     * @const {string} tmp
     */
    const tmp: string = '{prefix}$property must be an absolute directory path'

    return buildMessage(
      prefix => template(tmp, { prefix }),
      cast<Optional<ValidationOptions>>(get(args, 'constraints.0'))
    )()
  }

  /**
   * Check if `value` is a directory path.
   *
   * @public
   *
   * @param {unknown} value - Value to check
   * @return {value is string} `true` if value is directory path
   */
  public validate(value: unknown): value is string {
    return isString(value) && isAbsolute(value) && isDirectory(value)
  }
}

export default IsDirectoryConstraint
