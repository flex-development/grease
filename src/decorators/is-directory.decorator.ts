/**
 * @file Decorators - IsDirectory
 * @module grease/decorators/IsDirectory
 */

import type { PropertyDecorator } from '@flex-development/tutils'
import { ValidateBy, type ValidationOptions } from 'class-validator'
import validator from './is-directory.constraint'

/**
 * Check if a property value is a directory path.
 *
 * @decorator
 *
 * @param {ValidationOptions} [options] - Validation options
 * @return {PropertyDecorator} Property decorator
 */
const IsDirectory = (options: ValidationOptions = {}): PropertyDecorator => {
  return ValidateBy({
    async: validator.options.async,
    constraints: [options],
    name: validator.options.name,
    validator
  }, options)
}

export default IsDirectory
