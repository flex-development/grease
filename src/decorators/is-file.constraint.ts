/**
 * @file Decorators - IsFileConstraint
 * @module grease/decorators/IsFileConstraint
 */

import { isFile, toURL } from '@flex-development/mlly'
import {
  cast,
  get,
  isString,
  template,
  type Optional
} from '@flex-development/tutils'
import {
  ValidatorConstraint,
  buildMessage,
  type ValidatorConstraintInterface as IValidatorConstraint,
  type ValidationArguments,
  type ValidationOptions
} from 'class-validator'

/**
 * File validator.
 *
 * @class
 * @implements {IValidatorConstraint}
 */
@ValidatorConstraint(IsFileConstraint.options)
class IsFileConstraint implements IValidatorConstraint {
  /**
   * Validator constraint options.
   *
   * @public
   * @static
   * @readonly
   * @member {{ async: boolean; name: string }} options
   */
  public static readonly options: { async: boolean; name: string } = {
    async: false,
    name: isFile.name
  }

  /**
   * Get the default validation failure message.
   *
   * @see {@linkcode ValidationArguments}
   *
   * @public
   *
   * @param {ValidationArguments} args - Validation arguments
   * @return {string} Default validation failure message
   */
  public defaultMessage(args: ValidationArguments): string {
    return buildMessage(
      prefix => {
        return template('{prefix}$property must be a file path or url', {
          prefix
        })
      },
      cast<Optional<ValidationOptions>>(get(args, 'constraints.0'))
    )()
  }

  /**
   * Check if `value` is the path to a file.
   *
   * @public
   *
   * @param {unknown} value - Value to check
   * @return {value is string} `true` if value path to a file
   */
  public validate(value: unknown): value is string {
    return isString(value) && isFile(toURL(value))
  }
}

export default IsFileConstraint
