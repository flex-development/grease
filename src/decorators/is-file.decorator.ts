/**
 * @file Decorators - IsFile
 * @module grease/decorators/IsFile
 */

import type { PropertyDecorator } from '@flex-development/tutils'
import { ValidateBy, type ValidationOptions } from 'class-validator'
import validator from './is-file.constraint'

/**
 * Check if a property value is a file path or url.
 *
 * @decorator
 *
 * @param {ValidationOptions} [options] - Validation options
 * @return {PropertyDecorator} Property decorator
 */
const IsFile = (options: ValidationOptions = {}): PropertyDecorator => {
  return ValidateBy({
    async: validator.options.async,
    constraints: [options],
    name: validator.options.name,
    validator
  }, options)
}

export default IsFile
