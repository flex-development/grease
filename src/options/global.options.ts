/**
 * @file Options - GlobalOptions
 * @module grease/options/GlobalOptions
 */

import { IsDirectory } from '#src/decorators'
import pathe from '@flex-development/pathe'
import { DOT, get } from '@flex-development/tutils'
import { IsBoolean, IsString } from 'class-validator'

/**
 * Global grease options.
 *
 * @class
 */
class GlobalOptions {
  /**
   * Enable colorized output.
   *
   * @default true
   *
   * @public
   * @instance
   * @member {boolean} colors
   */
  @IsBoolean()
  public colors: boolean

  /**
   * Path to current working directory.
   *
   * @default pathe.resolve(DOT)
   *
   * @public
   * @instance
   * @member {string} cwd
   */
  @IsDirectory()
  public cwd: string

  /**
   * Enable verbose output.
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean} debug
   */
  @IsBoolean()
  public debug: boolean

  /**
   * Disable logs.
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean} silent
   */
  @IsBoolean()
  public silent: boolean

  /**
   * Tag prefix to consider when reading tags.
   *
   * @default ''
   *
   * @public
   * @instance
   * @member {string} tagprefix
   */
  @IsString()
  public tagprefix: string

  /**
   * Create a new options object.
   *
   * @param {Partial<GlobalOptions>?} [opts] - Global user options
   */
  constructor(opts?: Partial<GlobalOptions>) {
    this.colors = get(opts, 'colors', true)
    this.cwd = pathe.resolve(get(opts, 'cwd', DOT))
    this.debug = get(opts, 'debug', false)
    this.silent = get(opts, 'silent', false)
    this.tagprefix = get(opts, 'tagprefix', '')
  }
}

export default GlobalOptions
