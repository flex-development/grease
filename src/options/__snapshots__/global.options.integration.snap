// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:options/GlobalOptions > validation > should fail validation if schema is invalid 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isBoolean": "colors must be a boolean value",
    },
    "property": "colors",
    "value": 3,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isBoolean": "debug must be a boolean value",
    },
    "property": "debug",
    "value": 1,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isBoolean": "silent must be a boolean value",
    },
    "property": "silent",
    "value": 0,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "tagprefix must be a string",
    },
    "property": "tagprefix",
    "value": null,
  },
]
`;
